File: src/main/java/fp/yeyu/tos/SpiritEntity.java
package fp.yeyu.tos;

import io.github.yeyu.easing.EaseInImpl;
import io.github.yeyu.easing.EaseInOutImpl;
import io.github.yeyu.easing.function.InverseQuadratic;
import io.github.yeyu.easing.function.QuadraticFunction;
import io.github.yeyu.easing.interpolator.DoubleToDoubleInterpolator;
import io.github.yeyu.easing.player.PersistentFramefulEasePlayer;
import io.github.yeyu.easing.player.ReversingFramefulEasePlayer;
import net.fabricmc.api.EnvType;
import net.fabricmc.api.Environment;
import net.minecraft.entity.Entity;
import net.minecraft.entity.EntityType;
import net.minecraft.entity.EquipmentSlot;
import net.minecraft.entity.ItemEntity;
import net.minecraft.entity.damage.DamageSource;
import net.minecraft.entity.mob.MobEntity;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.item.ItemStack;
import net.minecraft.item.Items;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.particle.ParticleTypes;
import net.minecraft.util.Arm;
import net.minecraft.util.collection.DefaultedList;
import net.minecraft.util.math.Vec3d;
import net.minecraft.world.GameRules;
import net.minecraft.world.World;

import java.util.List;
import java.util.Optional;

public class SpiritEntity extends MobEntity {

    public static final DefaultedList<ItemStack> DEF = DefaultedList.ofSize(1, ItemStack.EMPTY);
    public static final double Y_MODIFIER_FACTOR = 0.5;
    private static final float MAX_DISTANCE_TO_TRACK_NEW = 20;
    protected ReversingFramefulEasePlayer<Double> posPlayer = null;
    protected static final int PARTICLE_TICK_DEF = 120;
    protected static final int PARTICLE_TICK_DEF_OFFSET = 10;
    protected static final double PARTICLE_LOWER_OFFSET = -1.5;
    protected static final double PARTICLE_UPPER_OFFSET = 2.5;
    protected static final double PARTICLE_OFFSET_RANGE = PARTICLE_UPPER_OFFSET - PARTICLE_LOWER_OFFSET;
    protected static final int TRACKING_TICK_DEF = 15;
    protected int particleTick = 5;
    protected int trackingTick = 5;
    protected Vec3d actualPosition = null;
    protected PlayerEntity tracking = null;


    protected boolean showSelf = false;
    protected boolean showHead = false;
    protected PlayerEntity closestPlayer = null;
    protected final int ALPHA_FRAMES = 35;

    private final PersistentFramefulEasePlayer<Double> selfAlphaLevelPlayer = new PersistentFramefulEasePlayer<>(
            new EaseInImpl<>(1d, 0d, QuadraticFunction.INSTANCE, DoubleToDoubleInterpolator.INSTANCE),
            ALPHA_FRAMES
    );

    public PersistentFramefulEasePlayer<Double> getSelfAlphaLevelPlayer() {
        return selfAlphaLevelPlayer;
    }

    public PersistentFramefulEasePlayer<Double> getHeadAlphaLevelPlayer() {
        return headAlphaLevelPlayer;
    }

    private final PersistentFramefulEasePlayer<Double> headAlphaLevelPlayer = new PersistentFramefulEasePlayer<>(
            new EaseInImpl<>(1d, 0d, QuadraticFunction.INSTANCE, DoubleToDoubleInterpolator.INSTANCE),
            ALPHA_FRAMES
    );

    protected SpiritEntity(EntityType<? extends MobEntity> entityType, World world) {
        super(entityType, world);
        setCanPickUpLoot(true);
    }

    @Override
    public boolean canImmediatelyDespawn(double distanceSquared) {
        return false;
    }

    @Override
    public void tick() {
        super.tick();
        if (world.isClient) return;
        particleTick = --particleTick % PARTICLE_TICK_DEF;
        if (particleTick == 0) {
            particleTick = PARTICLE_TICK_DEF + random.nextInt(PARTICLE_TICK_DEF_OFFSET);
            playAmbientParticle();
        }
    }

    private void updateTracking() {
        if (tracking == null || tracking.distanceTo(this) > MAX_DISTANCE_TO_TRACK_NEW) findNewPlayerToTrack();
    }

    private void findNewPlayerToTrack() {
        tracking = world.getClosestPlayer(this, MAX_DISTANCE_TO_TRACK_NEW);
    }

    @Override
    @Environment(EnvType.CLIENT)
    public void handleStatus(byte status) {
        if (status == 80) playAmbientParticle();
        else super.handleStatus(status);
    }

    private void playAmbientParticle() {
        if (world.isClient) {
            final int maxCount = random.nextInt(5);
            for (int i = 0; i < 1 + maxCount; ++i) {
                double x = getBlockPos().getX() + PARTICLE_LOWER_OFFSET + random.nextDouble() * PARTICLE_OFFSET_RANGE;
                double y = getBlockPos().getY() + PARTICLE_LOWER_OFFSET + random.nextDouble() * PARTICLE_OFFSET_RANGE;
                double z = getBlockPos().getZ() + PARTICLE_LOWER_OFFSET + random.nextDouble() * PARTICLE_OFFSET_RANGE;
                world.addParticle(ParticleTypes.DRAGON_BREATH, x, y, z, 0, 0.02, 0);
            }
        } else {
            world.sendEntityStatus(this, (byte) 80);
        }
    }

    @Override
    public void tickMovement() {

        this.world.getProfiler().push("looting");
        if (!this.world.isClient && this.canPickUpLoot() && this.isAlive() && !this.dead && this.world.getGameRules().getBoolean(GameRules.DO_MOB_GRIEFING)) {
            List<ItemEntity> list = this.world.getNonSpectatingEntities(ItemEntity.class, this.getBoundingBox().expand(1.0D, 0.0D, 1.0D));

            for (ItemEntity itemEntity : list) {
                if (!itemEntity.removed && !itemEntity.getStack().isEmpty() && !itemEntity.cannotPickup() && this.canGather(itemEntity.getStack())) {
                    this.loot(itemEntity);
                }
            }
        }
        this.world.getProfiler().swap("spirit_transparency");
        tickTransparency();
        this.world.getProfiler().swap("spirit_hover");
        if (posPlayer == null) configureEasePlayer();
        final double x = getPos().x;
        final double y = posPlayer.next();
        final double z = getPos().z;


        trackingTick = --trackingTick % TRACKING_TICK_DEF;
        if (trackingTick == 0) updateTracking();
        updatePosition(x, y, z);
        this.world.getProfiler().pop();
        if (tracking == null) return;
        this.world.getProfiler().push("spirit_look");
        lookAtEntity(tracking, 360f, 360f);
        this.world.getProfiler().pop();
    }

    private void tickTransparency() {
        closestPlayer = world.getClosestPlayer(this, 9);
        if (closestPlayer != null ^ showSelf) {
            showSelf = closestPlayer != null;
            selfAlphaLevelPlayer.setTransitionTo(showSelf ? 1d : 0d);
        }

        final PlayerEntity closestPlayerForHead = world.getClosestPlayer(this, 5);
        final boolean shouldShowHead = Optional.ofNullable(closestPlayerForHead).map(e -> {
            final int selectedSlot = e.inventory.selectedSlot;
            return e.inventory.getStack(selectedSlot).getItem();
        }).filter(e -> e == Items.LANTERN || e == Items.SOUL_LANTERN).isPresent();

        if (shouldShowHead != showHead) {
            showHead = shouldShowHead;
            headAlphaLevelPlayer.setTransitionTo(showHead ? 0.65 : 0.1);
        }

        if (showHead && random.nextDouble() < 0.05) {
            headAlphaLevelPlayer.setTransitionTo(random.nextBoolean() ? 0.65 : 0.75d);
        }
    }

    @Override
    public void lookAtEntity(Entity targetEntity, float maxYawChange, float maxPitchChange) {
        super.lookAtEntity(targetEntity, maxYawChange, maxPitchChange);
        this.headYaw = yaw;
    }

    private void configureEasePlayer() {
        posPlayer = new ReversingFramefulEasePlayer<>(
                new EaseInOutImpl<>(getPos().y,
                        getPos().y + Y_MODIFIER_FACTOR,
                        InverseQuadratic.INSTANCE,
                        QuadraticFunction.INSTANCE,
                        DoubleToDoubleInterpolator.INSTANCE),
                20 * 2);
        actualPosition = new Vec3d(getPos().x, getPos().y, getPos().z);
    }

    @Override
    public boolean isFallFlying() {
        return false;
    }

    @Override
    public void readCustomDataFromTag(CompoundTag tag) {
        super.readCustomDataFromTag(tag);
        if (tag.contains("actualPosition")) {
            final CompoundTag actualPosition = tag.getCompound("actualPosition");
            final double x = actualPosition.getDouble("x");
            final double y = actualPosition.getDouble("y");
            final double z = actualPosition.getDouble("z");
            setPos(x, y, z);
            posPlayer = null;
        }
    }

    @Override
    public void writeCustomDataToTag(CompoundTag tag) {
        super.writeCustomDataToTag(tag);
        final Vec3d pos = actualPosition == null ? getPos() : actualPosition;
        final CompoundTag compoundTag = new CompoundTag();
        compoundTag.putDouble("x", pos.x);
        compoundTag.putDouble("y", pos.y);
        compoundTag.putDouble("z", pos.z);
        tag.put("actualPosition", compoundTag);
    }

    @Override
    public boolean tryEquip(ItemStack equipment) {
        return true;
    }

    @Override
    public boolean damage(DamageSource source, float amount) {
        final Entity attacker = source.getAttacker();
        if (attacker == null) return super.damage(source, amount);
        attacker.damage(DamageSource.mob(this), amount);
        return super.damage(source, amount);
    }

    @Override
    public Iterable<ItemStack> getArmorItems() {
        return DEF;
    }

    @Override
    public ItemStack getEquippedStack(EquipmentSlot slot) {
        return DEF.get(0);
    }

    @Override
    public void equipStack(EquipmentSlot slot, ItemStack stack) {
    }

    @Override
    public Arm getMainArm() {
        return Arm.RIGHT;
    }



}

File: src/main/kotlin/fp/yeyu/tos/entity/SpiritEntity.kt
package fp.yeyu.tos.entity

import fp.yeyu.tos.TotemOfShadowEntry
import io.github.yeyu.easing.EaseInImpl
import io.github.yeyu.easing.EaseInOutImpl
import io.github.yeyu.easing.function.InverseQuadratic
import io.github.yeyu.easing.function.QuadraticFunction
import io.github.yeyu.easing.interpolator.DoubleToDoubleInterpolator
import io.github.yeyu.easing.player.PersistentFramefulEasePlayer
import io.github.yeyu.easing.player.ReversingFramefulEasePlayer
import net.fabricmc.api.EnvType
import net.fabricmc.api.Environment
import net.minecraft.entity.*
import net.minecraft.entity.ai.control.LookControl
import net.minecraft.entity.ai.control.MoveControl
import net.minecraft.entity.ai.goal.Goal
import net.minecraft.entity.ai.goal.LookAroundGoal
import net.minecraft.entity.attribute.DefaultAttributeContainer
import net.minecraft.entity.attribute.EntityAttributes
import net.minecraft.entity.damage.DamageSource
import net.minecraft.entity.mob.FlyingEntity
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.item.Items
import net.minecraft.particle.ParticleTypes
import net.minecraft.sound.SoundEvent
import net.minecraft.util.Arm
import net.minecraft.util.collection.DefaultedList
import net.minecraft.util.hit.BlockHitResult
import net.minecraft.util.hit.HitResult
import net.minecraft.util.math.Box
import net.minecraft.util.math.Vec3d
import net.minecraft.world.GameRules
import net.minecraft.world.RaycastContext
import net.minecraft.world.World
import java.util.*
import kotlin.random.asKotlinRandom

class SpiritEntity(entityType: EntityType<out FlyingEntity>?, world: World?) : FlyingEntity(entityType, world) {
    private val posPlayer = ReversingFramefulEasePlayer(
            EaseInOutImpl(-0.005,
                    0.005,
                    InverseQuadratic,
                    QuadraticFunction,
                    DoubleToDoubleInterpolator
            ),
            30)
    private var particleTick = 5

    private var showSelf = false
    private var showHead = false
    var closestPlayer: PlayerEntity? = null
    var lookingAtEntity: Entity? = null
    private val animationFrames = 35
    val selfAlphaLevelPlayer = PersistentFramefulEasePlayer(
            EaseInImpl(1.0, 0.0, QuadraticFunction, DoubleToDoubleInterpolator),
            animationFrames
    )

    val headAlphaLevelPlayer = PersistentFramefulEasePlayer(
            EaseInImpl(1.0, 0.1, QuadraticFunction, DoubleToDoubleInterpolator),
            animationFrames
    )

    override fun canImmediatelyDespawn(distanceSquared: Double): Boolean {
        return false
    }

    override fun initGoals() {
        super.initGoals()
        this.goalSelector.add(1, SpiritLookAtPlayerGoal(this))
        this.goalSelector.add(2, SpiritLookAroundGoal(this))
        this.goalSelector.add(3, SpiritFlyRandomlyGoal(this))
    }

    override fun tick() {
        super.tick()
        if (world.isClient) return
        particleTick = --particleTick % PARTICLE_TICK_DEF
        if (particleTick == 0) {
            particleTick = PARTICLE_TICK_DEF + random.nextInt(PARTICLE_TICK_DEF_OFFSET)
            playAmbientParticle()
        }
    }

    private fun hasPlayerAround(): Boolean = closestPlayer != null

    private fun isLookingAtSomeEntity(): Boolean =
            closestPlayer.let { it != null && squaredDistanceTo(it) < 81 } || lookingAtEntity.let { it != null && squaredDistanceTo(it) < 81 }

    @Environment(EnvType.CLIENT)
    override fun handleStatus(status: Byte) {
        if (status.toInt() == 80) playAmbientParticle() else super.handleStatus(status)
    }

    private fun playAmbientParticle() {
        if (world.isClient) {
            val maxCount = random.nextInt(5)
            for (i in 0 until 1 + maxCount) {
                val x = blockPos.x + PARTICLE_LOWER_OFFSET + random.nextDouble() * PARTICLE_OFFSET_RANGE
                val y = blockPos.y + PARTICLE_LOWER_OFFSET + random.nextDouble() * PARTICLE_OFFSET_RANGE
                val z = blockPos.z + PARTICLE_LOWER_OFFSET + random.nextDouble() * PARTICLE_OFFSET_RANGE
                world.addParticle(ParticleTypes.DRAGON_BREATH, x, y, z, 0.0, 0.02, 0.0)
            }
        } else {
            world.sendEntityStatus(this, 80.toByte())
        }
    }

    override fun tickMovement() {
        setNoGravity(true)
        super.tickMovement()
        world.profiler.push("looting")
        if (!world.isClient && canPickUpLoot() && this.isAlive && !dead && world.gameRules.getBoolean(GameRules.DO_MOB_GRIEFING)) {
            val list = world.getNonSpectatingEntities(ItemEntity::class.java, this.boundingBox.expand(2.0))
            for (itemEntity in list) {
                if (!itemEntity.removed && !itemEntity.stack.isEmpty && !itemEntity.cannotPickup() && canGather(itemEntity.stack)) {
                    loot(itemEntity)
                }
            }
        }
        world.profiler.pop()
        tickTransparency()
        tickIdle()
    }

    private fun rayCastTo(x: Double, y: Double, z: Double): BlockHitResult {
        return world.raycast(RaycastContext(pos, Vec3d(x, y, z), RaycastContext.ShapeType.OUTLINE, RaycastContext.FluidHandling.ANY, this))
    }

    private fun tickIdle() {
        world.profiler.push("spirit_hover")
        addVelocity(0.0, posPlayer.next(), 0.0)
        return world.profiler.pop()
    }

    private fun tickTransparency() {
        world.profiler.push("spirit_transparency")
        lookForClosestPlayer()
        val closestIsNull = closestPlayer == null
        if (closestIsNull != showSelf) {
            showSelf = closestIsNull
            selfAlphaLevelPlayer.transitionTo = if (showSelf) 1.0 else 0.0
        }
        val closestPlayerForHead = world.getClosestPlayer(this, 4.5)
        val shouldShowHead = Optional.ofNullable(closestPlayerForHead).map { e: PlayerEntity ->
            val selectedSlot = e.inventory.selectedSlot
            e.inventory.getStack(selectedSlot).item
        }.filter { e: Item -> e === Items.LANTERN || e === Items.SOUL_LANTERN }.isPresent
        if (shouldShowHead != showHead) {
            showHead = shouldShowHead
            headAlphaLevelPlayer.transitionTo = if (showHead) 0.65 else 0.1
        }
        if (showHead && random.nextDouble() < 0.05) {
            headAlphaLevelPlayer.transitionTo = if (random.nextBoolean()) 0.65 else 0.75
        }
        world.profiler.pop()
    }

    override fun isFallFlying(): Boolean = false

    override fun tryEquip(equipment: ItemStack): Boolean = true

    override fun damage(source: DamageSource, amount: Float): Boolean {
        source.attacker?.damage(DamageSource.mob(this), amount)
        return super.damage(source, amount)
    }

    override fun getArmorItems(): Iterable<ItemStack> {
        return DEF
    }

    override fun getEquippedStack(slot: EquipmentSlot): ItemStack {
        return DEF[0]
    }

    override fun equipStack(slot: EquipmentSlot, stack: ItemStack) {}

    override fun getMainArm(): Arm {
        return Arm.RIGHT
    }

    private fun lookForClosestPlayer() {
        closestPlayer = world.getClosestPlayer(this, 6.0)
    }

    private fun lookForAnyEntity() {
        lookingAtEntity = world.getNonSpectatingEntities(LivingEntity::class.java, Box(blockPos).expand(11.0)).getOrNull(1)
    }

    override fun getLookPitchSpeed(): Int = 360
    override fun getLookYawSpeed(): Int = 360

    override fun getAmbientSound(): SoundEvent = TotemOfShadowEntry.spiritEntityAmbientSound

    override fun getSoundPitch(): Float = 0.8f + random.nextFloat() * 0.4f

    companion object {
        private val DEF = DefaultedList.ofSize(1, ItemStack.EMPTY)
        private const val PARTICLE_TICK_DEF = 120
        private const val PARTICLE_TICK_DEF_OFFSET = 10
        private const val PARTICLE_LOWER_OFFSET = -1.5
        private const val PARTICLE_UPPER_OFFSET = 2.5
        private const val PARTICLE_OFFSET_RANGE = PARTICLE_UPPER_OFFSET - PARTICLE_LOWER_OFFSET

        fun createMobAttributes(): DefaultAttributeContainer.Builder {
            return createLivingAttributes()
                    .add(EntityAttributes.GENERIC_MOVEMENT_SPEED, 0.5)
                    .add(EntityAttributes.GENERIC_FOLLOW_RANGE, 16.0)
                    .add(EntityAttributes.GENERIC_ATTACK_KNOCKBACK)
        }
    }

    init {
        this.moveControl = SpiritMoveControl(this)
        this.lookControl = SpiritLookControl(this)
        setCanPickUpLoot(true)
    }

    class SpiritFlyRandomlyGoal(private val spiritEntity: SpiritEntity) : Goal() {
        override fun canStart(): Boolean {
            if (spiritEntity.hasPlayerAround()) return false
            val moveControl = spiritEntity.getMoveControl() as SpiritMoveControl
            return moveControl.movementTick > 200 || moveControl.movementTick == 0
        }

        override fun shouldContinue(): Boolean = false

        override fun canStop(): Boolean = true

        override fun start() {
            (spiritEntity.moveControl as SpiritMoveControl).moveToRandomTarget()
        }

        init {
            controls = EnumSet.of(Control.MOVE)
        }
    }

    class SpiritLookAtPlayerGoal(private val spiritEntity: SpiritEntity) : Goal() {

        override fun canStart(): Boolean {
            spiritEntity.lookForClosestPlayer()
            return !spiritEntity.getMoveControl().isMoving && hasPlayerAround()
        }

        private fun hasPlayerAround(): Boolean = spiritEntity.hasPlayerAround()

        override fun tick() {
            val lookingAt = spiritEntity.closestPlayer ?: return
            spiritEntity.lookControl.lookAt(lookingAt, 360f, 360f)
        }
    }

    class SpiritLookAroundGoal(private val spiritEntity: SpiritEntity) : LookAroundGoal(spiritEntity) {
        override fun canStart(): Boolean = !spiritEntity.getMoveControl().isMoving && !spiritEntity.hasPlayerAround() && super.canStart() && spiritEntity.random.nextDouble() < 0.3
        override fun shouldContinue(): Boolean = if (spiritEntity.isLookingAtSomeEntity()) false else super.shouldContinue()
        override fun canStop(): Boolean = spiritEntity.hasPlayerAround()

        override fun start() {
            spiritEntity.lookForAnyEntity()
            super.start()
        }

        override fun tick() {
            if (!spiritEntity.isLookingAtSomeEntity() || spiritEntity.lookingAtEntity == null) return super.tick()
            spiritEntity.lookControl.lookAt(spiritEntity.lookingAtEntity, 360f, 360f)
            spiritEntity.headYaw = spiritEntity.yaw
        }
    }

    class SpiritMoveControl(private val spiritEntity: SpiritEntity) : MoveControl(spiritEntity) {

        var movementTick: Int = 0

        override fun tick() {
            movementTick++

            if (state == State.MOVE_TO || state == State.JUMPING) {
                state = State.MOVE_TO
                super.tick()
                return
            }
        }

        fun moveToRandomTarget() {
            val movementRadius = 6.0
            val random = spiritEntity.random.asKotlinRandom()
            val x = spiritEntity.pos.x + random.nextDouble(-movementRadius, movementRadius)
            val y = spiritEntity.pos.y + random.nextDouble(-movementRadius, movementRadius)
            val z = spiritEntity.pos.z + random.nextDouble(-movementRadius, movementRadius)
            if (spiritEntity.rayCastTo(x, y, z).type != HitResult.Type.MISS) return
            moveTo(x, y, z, 1.0)
        }

        override fun moveTo(x: Double, y: Double, z: Double, speed: Double) {
            super.moveTo(x, y, z, speed)
            movementTick = 0
        }
    }

    class SpiritLookControl(spiritEntity: SpiritEntity) : LookControl(spiritEntity) {
        override fun tick() {
            yawSpeed = 360f
            pitchSpeed = 360f
            if (active) {
                active = false
                entity.headYaw = changeAngle(entity.headYaw, this.targetYaw, 360f)
                entity.pitch = changeAngle(entity.pitch, this.targetPitch, 360f)
            }
        }
    }
}

